src/data/fin/tuple/sort.lean

```diff
@@ -91,3 +91,84 @@ begin
 end
 
 end tuple
+
+namespace tuple
+
+open list
+
+variables {n : ℕ} {α : Type*}
+
+/-- If we swap two strictly decreasing entries in a tuple, then the result is lexicographically
+smaller than the original tuple. -/
+lemma lex_desc [preorder α] {f : fin n → α} {i j : fin n} (h₁ : i < j) (h₂ : f j < f i) :
+  to_lex (f ∘ equiv.swap i j) < to_lex f :=
+begin
+  simp only [has_lt.lt, pi.lex, pi.to_lex_apply, function.comp_app],
+  refine ⟨i, λ k (hik : k < i), _, _⟩,
+  { rw [equiv.swap_apply_of_ne_of_ne (ne_of_lt hik) (ne_of_lt $ hik.trans h₁)], },
+  { simpa only [equiv.swap_apply_left], }
+end
+
+/-- If a tuple `f` and a permutation of `f` are both monotone, then they are equal. -/
+lemma unique_monotone [partial_order α] {f : fin n → α} {σ : equiv.perm (fin n)} (hf : monotone f)
+  (hfσ : monotone (f ∘ σ)) : f ∘ σ = f :=
+of_fn_injective $ eq_of_perm_of_sorted (σ.of_fn_comp_perm f) hfσ.of_fn_sorted hf.of_fn_sorted
+
+variables [linear_order α] {f : fin n → α}
+
+/-- If a permutation of a tuple `f` is monotone, then it is equal to `f ∘ sort f`. -/
+lemma unique_sort' {σ : equiv.perm (fin n)} (h : monotone (f ∘ σ)) : f ∘ σ = f ∘ sort f :=
+begin
+  let σ' := (sort f)⁻¹ * σ,
+  have h' : f ∘ σ = (f ∘ sort f) ∘ σ',
+  { ext, simp only [function.comp_app, equiv.perm.coe_mul, equiv.perm.apply_inv_self], },
+  rw [h'],
+  refine unique_monotone (monotone_sort f) _,
+  rwa [← h'],
+end
+
+/-- If a tuple `f` is monotone, then it equals `f ∘ sort f`. -/
+lemma unique_sort (h : monotone f) : f = f ∘ sort f :=
+begin
+  have hf : f = f ∘ (1 : equiv.perm (fin n)),
+  { simp only [equiv.perm.coe_one, function.comp.right_id], },
+  conv_lhs {rw hf},
+  rw hf at h,
+  exact unique_sort' h,
+end
+
+/-- The sorted versions of a tuple `f` and any permutation of it agree. -/
+lemma sort_absorb {σ : equiv.perm (fin n)} : (f ∘ σ) ∘ (sort (f ∘ σ)) = f ∘ sort f :=
+begin
+  let τ := σ⁻¹ * (sort f),
+  have h' : (f ∘ σ) ∘ τ = f ∘ sort f,
+  { ext, simp only [equiv.perm.coe_mul, function.comp_app, equiv.perm.apply_inv_self] },
+  have hm : monotone ((f ∘ σ) ∘ τ) := by { rw [h'], exact monotone_sort _, },
+  exact (unique_sort' hm).symm.trans h',
+end
+
+/-- If a permutation `f ∘ σ` of the tuple `f` is not the same as `f ∘ sort f`, then `f ∘ σ`
+has a pair of strictly decreasing entries. -/
+lemma antitone_pair_of_not_sorted' {σ : equiv.perm (fin n)} (h : f ∘ σ ≠ f ∘ sort f) :
+  ∃ i j, i < j ∧ (f ∘ σ) j < (f ∘ σ) i :=
+begin
+  by_contra' hf,
+  have hm : monotone (f ∘ σ),
+  { intros i j hij,
+    cases eq_or_lt_of_le hij with heq hlt,
+    { rw [heq], },
+    { exact hf i j hlt, } },
+  exact h (unique_sort' hm),
+end
+
+/-- If the tuple `f` is not the same as `f ∘ sort f`, then it has a pair of strictly decreasing
+entries. -/
+lemma antitone_pair_of_not_sorted (h : f ≠ f ∘ sort f) : ∃ i j, i < j ∧ f j < f i :=
```

Comment:
Do we have the link between `f ≠ f ∘ sort f` and `not (monotone f)`?

Reply:
I only find `tuple.monotone_sort`, which says that `f ∘ sort f` is monotone. This does not purely formally imply that `f` is not monotone when `f ≠ f ∘ sort f`. This is why there is `tuple.unique_sort`, which is equivalent to `f ≠ f ∘ sort f → not (monotone f)` (line 131).

Reply:
`f ∘ sort f = f \iff monotone f` should be true, right? I think that would be a more powerful lemma to have.

Reply:
Good idea. If have replaced `unique_sort` by a new lemma `eq_comp_sort_iff_monotone` (with a one-line proof) and renamed `unique_sort'` to `sort_uniqe`.

Reply:
`sort_unique` could also be an iff, and it's more convenient if `unique_monotone` takes two permutations as arguments; see [my fork](https://github.com/leanprover-community/mathlib/compare/well_founded_induction_bot...alreadydone:mathlib:well_founded_induction_bot?expand=1#diff-3fca83b7393cf51b0fd014c0f357b654b27240852ff19ff09d3e5eae9afdd60aR127) (along with other golfs).